src/pipdeptree/_models/package.py

```diff
@@ -118,31 +111,40 @@ class DistPackage(Package):
 
     """
 
-    def __init__(self, obj: DistInfoDistribution, req: ReqPackage | None = None) -> None:
+    def __init__(self, obj: Distribution, req: ReqPackage | None = None) -> None:
         super().__init__(obj)
         self.req = req
-        self._project_name = ""
+        self.editable = False
+        self.direct_url = None
+        self.raw_name = self.project_name
 
     def requires(self) -> list[Requirement]:
-        return self._obj.requires()  # type: ignore[no-untyped-call,no-any-return]
+        req_list = []
+        req_name_list = []
+        if self._obj.requires:
+            for r in self._obj.requires:
+                req = Requirement(r)
+                if not req.marker and req.name not in req_name_list:
```

Comment:
This almost works, but we may miss direct dependencies that specify different markers (e.g. `python_version`). I am not seeing any easy way of finding what marker it is besides converting it into a string:
```console
>>> r = Requirement("what; extra == 'test'")
>>> str(r.marker)
'extra == "test"'
```

From reading the [docs](https://packaging.python.org/en/latest/specifications/dependency-specifiers/#environment-markers), I think it is safe to keep dependencies that have any marker besides "extra".

Reply:
We could maybe do something like:
```suggestion
               is_extra_req = req.marker and str(req.marker).startswith("extra")
                if not is_extra_req and req.name not in req_name_list:
```

Reply:
After checking some dists, I found that the `dist.requires` may return something like `"importlib-metadata>=6; python_version < '3.8' and extra == 'testing'"`. This situation can not be caught by `startswith`.
So I think maybe we can search for `[blank]extra ==`. I checked several pkgs and it looks right.
I will check the generating of `Requires-Dist` format in METADATA file when I have time.
ref: https://github.com/python/cpython/blob/3.12/Lib/importlib/metadata/__init__.py#L559

Reply:
Yes, nice catch; I forgot to account for multiple markers being used.

> [blank]extra[blank]==

I think this would work too, but if by [blank] you mean a single whitespace character then it might be slightly more complex as it looks like parsers can accept 0-to-many whitespace characters around markers:
```
marker_op     = version_cmp | (wsp* 'in') | (wsp* 'not' wsp+ 'in')
marker_var    = wsp* (env_var | python_str)
marker_expr   = marker_var marker_op marker_var
              | wsp* '(' marker wsp* ')'
marker_and    = marker_expr wsp* 'and' marker_expr
              | marker_expr
marker_or     = marker_and wsp* 'or' marker_and
                  | marker_and
marker        = marker_or
```
(see [here](https://packaging.python.org/en/latest/specifications/dependency-specifiers/#grammar) and ([here](https://packaging.python.org/en/latest/specifications/dependency-specifiers/#whitespace)) for more details)

Reply:
I went snooping around in `packaging`'s GitHub repo to see if we can use anything else to solve this problem; I think our approach is the best option as all I can find is a [feature request](https://github.com/pypa/packaging/issues/496) that talks about exposing an API for the AST that is built.

Reply:
Oh, it seems that someone has the same needs as us, but it has not been responded. So, let's just use our own approach  for now.

Reply:
> I think this would work too, but if by [blank] you mean a single whitespace character then it might be slightly more complex as it looks like parsers can accept 0-to-many whitespace characters around markers:

You are right. I think we can `strip` the marker before handling it, for getting rid of the whitespace.

Reply:
Sorry, I made a mistake. `python_version < "3.8" and extra == "testing"` cannot be stripped, because it will turn into a string hardly be parsed.

Reply:
I think we can solve it with a regular expression.
```
re.search(r'\bextra\s*==', s)
```

Reply:
I think this is good for now; we'll add unit tests for `requires()` as this PR get closer to solving the issue.